control-plane/config-entries/controllers/configentry_controller.go

```diff
@@ -38,8 +38,13 @@ const (
 // Controller is implemented by CRD-specific controllers. It is used by
 // ConfigEntryController to abstract CRD-specific controllers.
 type Controller interface {
-	// Update updates the state of the whole object.
-	Update(context.Context, client.Object, ...client.UpdateOption) error
+	// AddFinalizersPatch creates a patch with the original finalizers with new ones appended to the end.
+	AddFinalizersPatch(obj client.Object, finalizers ...string) *FinalizerPatch
+	// Remove
+	RemoveFinalizersPatch(obj client.Object, finalizers ...string) *FinalizerPatch
+	// Patch patches the object. This should only ever be used for updating the metadata of an object, and not object
+	// spec or status. Updating the spec could have unintended consequences like defaulting zero values.
+	Patch(ctx context.Context, obj client.Object, patch client.Patch, opts ...client.PatchOption) error
```

Comment:
@hashi-derek does this interface look ok? I tried keeping it as AddFinalizers and RemoveFinalizers, implemented by the FinalizerPatcher in finalizer_patch.go but if I do the patching within those functions, I'd need to pass in the k8s client to to the FinalizerPatcher. Keeping Patch as a function on the controller allows the FinalizerPatcher to not need another k8s client that's already being embedded in the controller implementations.

Reply:
Yeah, this is fine. It's simpler than having the client passed in on each function call.